src/graft_dag.cpp

```diff
@@ -0,0 +1,418 @@
+
+// ** Comparators
+
+int GraftDAG::Compare(const GraftDAG &other) const {
+  return GraftDAG::Compare(*this, other);
+}
+
+int GraftDAG::Compare(const GraftDAG &lhs, const GraftDAG &rhs) {
+  const SubsplitDAG &lhs_host = lhs.GetHostDAG();
+  const SubsplitDAG &rhs_host = rhs.GetHostDAG();
+  // (1) Compare host DAGs.
+  int host_diff = SubsplitDAG::Compare(lhs_host, rhs_host);
+  if (host_diff != 0) {
+    return host_diff;
+  }
+  // (2) Compare graft nodes.
+  BitsetVector lhs_nodes = lhs.GetSortedVectorOfNodeBitsets();
+  BitsetVector rhs_nodes = rhs.GetSortedVectorOfNodeBitsets();
+  if (lhs_nodes != rhs_nodes) {
+    return (lhs_nodes < rhs_nodes) ? -1 : 1;
+  }
+  // (3) Compare graft edges.
+  BitsetVector lhs_edges = lhs.GetSortedVectorOfEdgeBitsets();
+  BitsetVector rhs_edges = rhs.GetSortedVectorOfEdgeBitsets();
+  if (lhs_edges != rhs_edges) {
+    return (lhs_edges < rhs_edges) ? -1 : 1;
+  }
+  return 0;
+}
+
+int GraftDAG::CompareToDAG(const SubsplitDAG &other) const {
+  return GraftDAG::CompareToDAG(*this, other);
+}
+
+int GraftDAG::CompareToDAG(const GraftDAG &lhs, const SubsplitDAG &rhs) {
+  // Compare taxon counts.
+  int taxon_diff = lhs.TaxonCount() - rhs.TaxonCount();
+  if (taxon_diff != 0) {
+    return taxon_diff;
+  }
+  // Compare nodes.
+  BitsetVector lhs_nodes = lhs.GetSortedVectorOfNodeBitsets();
+  BitsetVector rhs_nodes = rhs.GetSortedVectorOfNodeBitsets();
+  if (lhs_nodes != rhs_nodes) {
+    return (lhs_nodes < rhs_nodes) ? -1 : 1;
+  }
+  // Compare edges.
+  BitsetVector lhs_edges = lhs.GetSortedVectorOfEdgeBitsets();
+  BitsetVector rhs_edges = rhs.GetSortedVectorOfEdgeBitsets();
+  if (lhs_edges != rhs_edges) {
+    return (lhs_edges < rhs_edges) ? -1 : 1;
+  }
+  return 0;
+}
+
+// ** Modify GraftDAG
+
+void GraftDAG::CreateGraftNode(const Bitset &node_subsplit) {
+  // Do not add node to DAG if already exists in DAG.
+  if (ContainsNode(node_subsplit)) {
+    return;
+  }
+  // else, add node to graft node list.
+  size_t node_id = NodeCount();
+  storage_.AddVertex({node_id, node_subsplit}, HostNodeCount());
+  // insert node into subsplit map.
+  SafeInsert(subsplit_to_id_, node_subsplit, node_id);
+}
+
+void GraftDAG::CreateGraftEdge(const size_t parent_id, const size_t child_id) {
+  // find relationship between nodes.
+  Bitset parent_subsplit = GetDAGNode(parent_id).GetBitset();
+  Bitset child_subsplit = GetDAGNode(child_id).GetBitset();
+  bool is_left =
+      !Bitset::SubsplitIsChildOfWhichParentClade(parent_subsplit, child_subsplit);
+  CreateGraftEdge(parent_id, child_id, is_left);
+}
+
+void GraftDAG::CreateGraftEdge(const size_t parent_id, const size_t child_id,
+                               const bool is_left) {
+  // Add edge to edge list.
+  Assert(ContainsNode(parent_id), "Node with the given parent_id does not exist.");
+  Assert(ContainsNode(child_id), "Node with the given child_id does not exist.");
+
+  auto parent_node = GetDAGNode(parent_id);
```

Comment:
a lot of variables here can be marked const... perhaps all of them? Let's just get in the habit of making that the default.

Reply:
Okay, I'll make a pass to see if any other can be updated.